website/docs/topics/non-openai-models/cloud-anthropic.ipynb

```diff
@@ -132,39 +143,79 @@
+    "            return [None]\n",
+    "        res = []\n",
+    "        if TOOL_ENABLED:\n",
+    "            for choice in messages:\n",
+    "                if choice.type == \"tool_use\":\n",
+    "                    res.insert(0, self.response_to_openai_message(choice))\n",
+    "                    self._last_tooluse_status[\"tool_use\"] = choice.model_dump()\n",
+    "                else:\n",
+    "                    res.append(choice.text)\n",
+    "                    self._last_tooluse_status[\"think\"] = choice.text\n",
+    "\n",
+    "            return res\n",
     "\n",
-    "        # claude python SDK and API not yet support function calls\n",
+    "        else:\n",
+    "            return [  # type: ignore [return-value]\n",
+    "                choice.text if choice.message.function_call is not None else choice.message.content  # type: ignore [union-attr]\n",
+    "                for choice in messages\n",
+    "            ]\n",
     "\n",
     "    def create(self, params: Dict[str, Any]) -> Completion:\n",
-    "        \"\"\"Create a completion for a given config using openai's client.\n",
+    "        \"\"\"Create a completion for a given config.\n",
     "\n",
     "        Args:\n",
-    "            client: The openai client.\n",
     "            params: The params for the completion.\n",
     "\n",
     "        Returns:\n",
     "            The completion.\n",
     "        \"\"\"\n",
-    "        if \"messages\" in params:\n",
-    "            raw_contents = params[\"messages\"]\n",
-    "            if raw_contents[0][\"role\"] == \"system\":\n",
-    "                system_message = raw_contents[0][\"content\"]\n",
-    "                raw_contents = raw_contents[1:]\n",
-    "                params[\"messages\"] = raw_contents\n",
-    "                params[\"system\"] = system_message\n",
-    "            completions: Completion = self._client.messages  # type: ignore [attr-defined]\n",
+    "        if \"tools\" in params:\n",
+    "            converted_functions = self.convert_tools_to_functions(params[\"tools\"])\n",
+    "            params[\"functions\"] = params.get(\"functions\", []) + converted_functions\n",
+    "\n",
+    "        raw_contents = params[\"messages\"]\n",
+    "        processed_messages = []\n",
+    "        for message in raw_contents:\n",
+    "\n",
+    "            if message[\"role\"] == \"system\":\n",
+    "                params[\"system\"] = message[\"content\"]\n",
+    "            elif message[\"role\"] == \"function\":\n",
+    "                processed_messages.append(self.return_function_call_result(message[\"content\"]))\n",
+    "            elif \"function_call\" in message:\n",
+    "                processed_messages.append(self.restore_last_tooluse_status())\n",
+    "            elif message[\"content\"] == \"\":\n",
+    "                # I'm not sure how to elegantly terminate the conversation, please give me some advice about this.\n",
```

Comment:
I believe just replying "TERMINATE" here should do the trick

Reply:
Although come to think of it if there is a different `is_terminate` function specified, then it won't work. Is there a handle on that? maybe the response could take that into account

Reply:
Thanks for the suggestion, but I think an instruct may perform better than just `TERMINATE`. And unfortunately I haven't find such a handle as I explored.